### PR TITLE
feat: device syncing

### LINKED ISSUE #5837: Sync profile on restore seed phrase onboarding workflow
### Description

see figma:
- https://www.figma.com/file/17fc13UBFvInrLgNUKJJg5/Kuba%E2%8E%9CDesktop?node-id=1023%3A298171
- https://www.figma.com/file/idUoxN7OIW2Jpp3PMJ1Rl8/%E2%9A%99%EF%B8%8F-Settings-%7C-Desktop?node-id=1592%3A128606

Support syncing from desktop, if mobile does not support this yet, then it should be done in a different task.

Screen Shot 2022-05-30 at 7 08 40 PM

goals:
- [ ] write feature spec describing syncing in https://github.com/status-im/feature-specs
- [ ] update settings syncing screen
- [ ] Support syncing with another desktop app

note:
- status-go apis for this should already exist, but it's possible some might need to be implemented
- keycard syncing is out of scope
- [PR#5756](https://github.com/status-im/status-desktop/pull/5756) might be relevant if scanning QR codes is required

### PR DESCRIPTION
**Pull request by igor-sirotin**
Closes #5837
Closes https://github.com/status-im/status-desktop/issues/7861

Requires:
- https://github.com/status-im/status-go/pull/3184
- https://github.com/status-im/nim-status-go/pull/22
- https://github.com/status-im/status-go/pull/3228

# Sync 💻 → 💻 

https://user-images.githubusercontent.com/25482501/218459894-043f2118-0ac9-4f45-a484-ee58539a5e91.mov

Notes:
- ~~After closing the first app instance and restarting the second one, it's possible to login to the synced account~~
- ~~If hard-code other port (60003) on second instance, it logins automatically with no problems~~

# Sync 📱 ←→ 💻  

✅ UI is implemented
✅ The sync works
❌ Fail to login with _Invalid password_

Fixes are on the go:
- https://github.com/status-im/status-desktop/pull/9528 
- https://github.com/status-im/status-go/pull/3179

### CODE REVIEW

**Comment by igor-sirotin** [comment]
@caybro, I will squash them on PR merging, we recently have enabled that option in repo settings 🙂

**Comment by caybro** [comment]
> @caybro, I will squash them on PR merging, we recently have enabled that option in repo settings slightly_smiling_face

Sure thing but pls rebase first :)

**Comment by richard-ramos** [comment]
Squashed all the commits and rebased the PR because it was very annoying to do the rebase with all the commits separate :cry:

File: ui/StatusQ/src/StatusQ/Core/Utils/Utils.qml [hunk 1]
```diff
@@ -266,6 +266,11 @@ QtObject {
     function encodeUtf8(str){
         return unescape(encodeURIComponent(str));
     }
+
+    function deviceIcon(deviceType) {
+        const isMobileDevice = root.deviceType === "ios" || root.deviceType === "android"
```
[the remark this answers is not in the document]
Comment (caybro) [reply ui/StatusQ/src/StatusQ/Core/Utils/Utils.qml:271]: There is no `root.deviceType` here

File: ui/StatusQ/src/StatusQ/Components/StatusSyncDeviceDelegate.qml [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import QtQuick 2.14
+import QtQuick.Controls 2.13
+import QtQuick.Layouts 1.13
+import QtGraphicalEffects 1.13
```
Comment (caybro) [inline ui/StatusQ/src/StatusQ/Components/StatusSyncDeviceDelegate.qml:4]: Needed here?

File: ui/imports/shared/controls/GetSyncCodeDesktopInstructions.qml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    RowLayout {
+        height: d.listItemHeight
+
+        StatusBaseText {
+            Layout.alignment: Qt.AlignVCenter
+            font.pixelSize: 15
+            color: Theme.palette.baseColor1
+            text: qsTr("%2Open").arg("2. ")
```
Comment (caybro) [inline ui/imports/shared/controls/GetSyncCodeDesktopInstructions.qml:37]: Also in the rest of this file

File: ui/imports/shared/controls/GetSyncCodeMobileInstructions.qml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    RowLayout {
+        height: d.listItemHeight
+
+        StatusBaseText {
+            Layout.alignment: Qt.AlignVCenter
+            font.pixelSize: 15
+            color: Theme.palette.baseColor1
+            text: qsTr("%1Open Status App on your mobile device").arg("1. ")
```
Comment (caybro) [inline ui/imports/shared/controls/GetSyncCodeMobileInstructions.qml:26]: And here

File: ui/imports/shared/popups/GetSyncCodeInstructionsPopup.qml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    id: root
+
+    enum Source {
+        Mobile,
+        Desktop
+    }
+
+    title: qsTr("How to get a sync code on...")
+    padding: 40
+    footer: Item { }
```
Comment (caybro) [inline ui/imports/shared/popups/GetSyncCodeInstructionsPopup.qml:19]: `null`

**Review by caybro — CHANGES_REQUESTED** [review]
Some relatively minor things inline, otherwise great job!

**Review by caybro — APPROVED** [review]
Code LGTM, will need some heavy testing :) This is a big change

File: src/app_service/service/devices/service.nim [hunk 6]
```diff
@@ -1,66 +1,100 @@
...
+const SIGNAL_LOCAL_PAIRING_EVENT* = "localPairingEvent"
+const SIGNAL_LOCAL_PAIRING_STATUS_UPDATE* = "localPairingStatusUpdate"
 
 QtObject:
   type Service* = ref object of QObject
     events: EventEmitter
+    threadpool: ThreadPool
     settingsService: settings_service.Service
+    accountsService: accounts_service.Service
+    localPairingStatus: LocalPairingStatus
```
Comment (saledjenic) [inline src/app_service/service/devices/service.nim:46]: @igor-sirotin wouldn't be better if we could do a lazy initialization here? (having this `nil` at the beginning and initialize it only if there is a need for that)

File: ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+
+     footer: StatusDialogFooter {
+        rightButtons: ObjectModel {
+            StatusButton {
+                text: qsTr("Done")
+                enabled: nameInput.text !== ""
+                onClicked : {
+                    root.devicesStore.setName(nameInput.text.trim())
```
Comment (saledjenic) [inline ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml:53]: @igor-sirotin I am not able to set name for any device.

Comment (igor-sirotin) [reply ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml:53]: Yeah, this is something I saw at last moment 🙁 
Probably easier to fix it as a separate PR, it's kinda unrelated issue.

Comment (richard-ramos) [reply ui/app/AppLayouts/Profile/popups/SyncDeviceCustomizationPopup.qml:53]: Tracked here: https://github.com/status-im/status-desktop/issues/9584

**Review by richard-ramos — APPROVED** [review]

**Review by saledjenic — APPROVED** [review]

**Comment by anastasiyaig** [comment]
@igor-sirotin hi, could u please clarify what is implemented there in scope of this PR? I am interested in screens (which ones from https://www.figma.com/file/idUoxN7OIW2Jpp3PMJ1Rl8/%E2%9A%99%EF%B8%8F-Settings-%7C-Desktop?node-id=1592%3A128590&t=D0V2ac3MQa6UDKHq-4 ) and which syncing functionality from [here](https://docs.google.com/spreadsheets/d/1VyVIg5ZaVSlKWZRHSLzSMCMGJsbprkWUfEqCUJKma8w/edit?pli=1#gid=507125679) (based on the branch name it should be Profile, but would be cool to clarify)

@caybro @richard-ramos please bring some light if u have any information :)

**Comment by jrainville** [comment]
@anastasiyaig sorry, I'm not sure. Maybe it's just what's seen in the video?

@richard-ramos might know more

**Comment by richard-ramos** [comment]
Should be what's seen in the video. The figma file includes some QR scanning screens but that should be tested in this PR: https://github.com/status-im/status-desktop/pull/9464 (it's still in progress)

**Comment by anastasiyaig** [comment]
@igor-sirotin @richard-ramos sadly i was not able to reproduce even the flow described on the screen with any seed phrases (that were created before, not within this PR or new ones). It is crashing all the time, i cant login to second instance. What i did:

1. on instance 1 i recover known seed (for example the one I use for dogfooding)
2. on instance 1 i get the sync code
3. on instance 2 i paste the sync code
4. as result, the second instance is always crashing. When restarted - there is no sign of recovered account, i am starting from scratch.

```
Traceback (most recent call last)
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/core/tasks/threadpool.nim(113) taskThread
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-task-runner/vendor/nim-chronos/chronos/asyncloop.nim(1082) waitFor
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-task-runner/vendor/nim-chronos/chronos/asyncloop.nim(279) poll
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/core/tasks/threadpool.nim(98) runner
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app_service/service/devices/async_tasks.nim(11) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-status-go/status_go.nim(379) inputConnectionStringForBootstrapping
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Traceback (most recent call last)
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/nim_status_client.nim(180) nim_status_client
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/nim_status_client.nim(177) mainProc
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimqml/src/nimqml/private/qguiapplication.nim(29) exec
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimqml/src/nimqml/private/qobject.nim(49) qobjectCallback
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimqml/src/nimqml/private/qobject.nim(34) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(569) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/modules/startup/view.nim(101) onPrimaryActionClicked
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/modules/startup/io_interface.nim(46) onPrimaryActionClicked
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/modules/startup/module.nim(178) onPrimaryActionClicked
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/modules/startup/view.nim(83) setCurrentStartupState
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/modules/startup/internal/state_wrapper.nim(19) setStateObj
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(564) stateWrapperChanged
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nimqml/src/nimqml/private/qobject.nim(32) emit
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by igor-sirotin** [comment]
@anastasiyaig, what's in the video is mainly what should be checked.

### List of things to check:

1. Setup Desktop with another Desktop
2. Setup Desktop with Mobile
3. Setup Mobile with Desktop
4. Updated `Syncing` settings page design

### What won't work

- Please, check all notes in PR description
- mobile<->desktop should already probably work after merging https://github.com/status-im/status-desktop/pull/9528. But I didn't check that. 
@richard-ramos, can you please make sure it's rebase on top of it?

### Syncing steps instructions:

| Step | Instance A | Instance B |
|------|------------|------------|
| 1 | Create a new account |  |
| 2 | Go to settings -> Syncing |  |
| 3 | Click "Setup syncing" | |
| 4 | | Run onboarding with "Setup with a sync code" |
| 5 | Copy the sync code |  |
| 6 | | Paste the sync code |
| 7 | Note the popup shows syncing | Note the app auto-logins with same account |

**Comment by igor-sirotin** [comment]
@anastasiyaig, it seems to be getting some error on `inputConnectionStringForBootstrapping`. 
@richard-ramos, can you please check it out? 🙄

**Comment by richard-ramos** [comment]
Sure! taking a look at it now.

**Comment by anastasiyaig** [comment]
@richard-ramos thank you! Also guys, is there any limitation how many devices can be synced at the same time? @igor-sirotin do u know? :) @John-44

**Comment by richard-ramos** [comment]
Looks like it's crashing in status-go according to gdb:
```go
#0  0x00007ffff2abbe4a in github.com/status-im/status-go/server/pairing.NewPayloadSourceForClient.func1 (conf=0x1c0005461e0, ~r0=<optimized out>, ~r1=...)
    at /home/richard/status/status-desktop/vendor/status-go/server/pairing/payload_manager.go:86
#1  0x00007ffff2abc0e2 in github.com/status-im/status-go/server/pairing.unmarshalPayloadSourceConfig (configJSON=..., 
    successCallback={void (github.com/status-im/status-go/server/pairing.PayloadSourceConfig *, github.com/status-im/status-go/server/pairing.PayloadSourceConfig **, error *)} 0x1c000de1ca0, ~r0=<optimized out>, ~r1=...)
    at /home/richard/status/status-desktop/vendor/status-go/server/pairing/payload_manager.go:121
#2  0x00007ffff2abbdda in github.com/status-im/status-go/server/pairing.NewPayloadSourceForClient (configJSON=..., mode=<optimized out>, ~r0=<optimized out>, ~r1=...)
    at /home/richard/status/status-desktop/vendor/status-go/server/pairing/payload_manager.go:85
```

I'll attempt to fix it

**Comment by richard-ramos** [comment]
Sorry for the delay @anastasiyaig .
In https://github.com/status-im/status-go/pull/3179 some breaking changes were introduced and not present in this PR, so when I rebased it, it ended up not working.

@igor-sirotin FYI we have to track this issue too https://github.com/status-im/status-go/issues/3198, since there will be some refactoring on status-go in the future!

**Comment by Samyoul** [comment]
@igor-sirotin @richard-ramos with regards to https://github.com/status-im/status-go/issues/3198 I will detail specs of the changes so that you are very aware of what the latest requirements are. I'm sorry for the need to make breaking changes at this late stage.

**Comment by anastasiyaig** [comment]
thank you @richard-ramos ! It works now :) 2 questions:

1. the profile details are actually not syncing when i change them on any instance, lets say instance 1 - i wonder how often we re-sync that (if we even do) or is / will it be possible to trigger manual sync update across devices?
2. second instance of the app crashes after the device is found (works after relogin):

```
Warning: qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27:5: QML Image: Cannot open: qrc:/StatusQ/src/StatusQ/Components/%3Cuser image%3E (qrc:/StatusQ/src/StatusQ/Components/StatusRoundedImage.qml:27, )
WRN 2023-02-23 15:49:38.034+03:00 Error decoding signal                      topics="signals-manager" tid=8583047 file=signals_manager.nim:43 err="Unknown signal received: mediaserver.started"
Traceback (most recent call last)
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/core/tasks/threadpool.nim(113) taskThread
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-task-runner/vendor/nim-chronos/chronos/asyncloop.nim(1082) waitFor
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-task-runner/vendor/nim-chronos/chronos/asyncloop.nim(279) poll
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app/core/tasks/threadpool.nim(98) runner
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/src/app_service/service/devices/async_tasks.nim(11) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482/vendor/nim-status-go/status_go.nim(379) inputConnectionStringForBootstrapping
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Warning: QObject::~QObject: Timers cannot be stopped from another thread (:0, )
INF 2023-02-23 15:49:41.516+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-01.ac-cn-hongkong-c.status.prod nodeAddress=/dns4/node-01.ac-cn-hongkong-c.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAkvEZgh3KLwhLwXg95e5ojM8XykJ4Kxi2T7hk22rnA7pJC
INF 2023-02-23 15:49:41.516+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-01.gc-us-central1-a.status.prod nodeAddress=/dns4/node-01.gc-us-central1-a.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAkwBp8T6G77kQXSNMnxgaMky1JeyML5yqoTHRM8dbeCBNb
INF 2023-02-23 15:49:41.516+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-01.do-ams3.status.prod nodeAddress=/dns4/node-01.do-ams3.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAm6HZZr7aToTvEBPpiys4UxajCTU97zj5v7RNR2gbniy1D
INF 2023-02-23 15:49:41.517+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-02.ac-cn-hongkong-c.status.prod nodeAddress=/dns4/node-02.ac-cn-hongkong-c.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAmFy8BrJhCEmCYrUfBdSNkrPw6VHExtv4rRp1DSBnCPgx8
INF 2023-02-23 15:49:41.517+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-02.do-ams3.status.prod nodeAddress=/dns4/node-02.do-ams3.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAmSve7tR5YZugpskMv2dmJAsMUKmfWYEKRXNUxRaTCnsXV
INF 2023-02-23 15:49:41.517+03:00 initMailservers                            topics="mailservers-service" tid=8583047 file=service.nim:202 name=node-02.gc-us-central1-a.status.prod nodeAddress=/dns4/node-02.gc-us-central1-a.status.prod.statusim.net/tcp/30303/p2p/16Uiu2HAmDQugwDHM3YeUp86iGjrUvbdw3JPRgikC7YoGBsT2ymMg
INF 2023-02-23 15:49:41.519+03:00 fetch mailservers                          topics="mailservers-service" tid=8583047 file=service.nim:208 rpc_proc=mailservers_getMailservers response="(jsonrpc: \"2.0\", result: null, id: 0, error: nil)"
```
3. when being on `device found` screen, the display name is set as <display name> (same as on the video). Perhaps will be implemented later?
4. why i am having so many devices available to sync in the profile? :D 

Screenshot 2023-02-23 at 15 59 02

**Comment by richard-ramos** [comment]
I'll check what's up with all these issues! 

> why i am having so many devices available to sync in the profile? :D

You can never have enough devices ;) .   (Will check this too)

**Comment by richard-ramos** [comment]
> the profile details are actually not syncing when i change them on any instance, lets say instance 1 - i wonder how often we re-sync that (if we even do) or is / will it be possible to trigger manual sync update across devices?

Looks like only the display name and the bio are being synced, and not the social links. cc-ing: @Samyoul and @qfrank to confirm!

**Comment by richard-ramos** [comment]
> why i am having so many devices available to sync in the profile? :D

Trying unsuccesfully to reproduce this. Did you try syncing with all these devices and the same account? or only tried to sync one device and got them duplicated?

**Comment by Samyoul** [comment]
> > the profile details are actually not syncing when i change them on any instance, lets say instance 1 - i wonder how often we re-sync that (if we even do) or is / will it be possible to trigger manual sync update across devices?
> 
> @anastasiyaig: Looks like only the display name and the bio are being synced, and not the social links. cc-ing: @Samyoul and @qfrank to confirm!

This seems to be 2 things:

1 ) The initial syncing of related identity information (social links)

2) Keeping the 2 installations in sync after local pairing.

This is something we are aware of. At the moment only an initial sync is triggered. @qfrank is working on implementing the actual linking the 2 devices together, the actual pairing of devices, https://github.com/status-im/status-go/pull/3202

So at the moment, after the initial bootstrapping of a device the 2 installations don't talk to each other. We are resolving that.

**Comment by Samyoul** [comment]
> why i am having so many devices available to sync in the profile? :D

@anastasiyaig

This looks like the test installation is maintaining a record of all the instances of other installations it has seen in the past. This could be the result of using the same installation for multiple pairing tests.

Every time you create a new installation a message is fired informing all devices using the same private key that there is a new installation. The device stores this data whether you pair with those devices or not.

**Comment by anastasiyaig** [comment]
@Samyoul thank u very much for clarification!

@igor-sirotin or @richard-ramos , can we please do a quick UI fix for this screen (it should not contain `From` word in the tabs names per design)

**Design:**
image

**PR:**
Screenshot 2023-02-23 at 16 39 10

**Comment by richard-ramos** [comment]
@igor-sirotin @caybro 
I introduced some new changes in last commit, that were needed to be able to display the username, profile picture, color and hash (with this new info being provided by status-go's https://github.com/status-im/status-go/pull/3228). If you have some time, please re-review, specially the QML changes.
Thanks in advance!

File: ui/imports/shared/views/DeviceSyncingView.qml [hunk 8]
```diff
@@ -39,15 +41,39 @@ Item {
...
+            active: root.userPublicKey == ""
+            Layout.alignment: Qt.AlignHCenter
+            sourceComponent: UserImage {
+                name: root.userDisplayName
+                colorId: root.userColorId
+                colorHash: root.userColorHash
+                image: root.userImage
+                interactive: false
+                imageWidth: 80
+                imageHeight: 80
```
Comment (caybro) [inline ui/imports/shared/views/DeviceSyncingView.qml:75]: You could use the same `Loader` and the same `sourceComponent`; just set the different properties based on `root.userPublicKey`

**Review by caybro — APPROVED** [review]
Still LGTM, one suggestion though for the QML part

**Comment by anastasiyaig** [comment]
@richard-ramos will u be able to address some issues i listed in https://github.com/status-im/status-desktop/pull/9482#issuecomment-1442138760 ? Or i should wait for Igor?

**Comment by richard-ramos** [comment]
@anastasiyaig I'll fix them tomorrow (I'm OoO today!)

**Comment by richard-ramos** [comment]
@anastasiyaig I did a fix for the UI issues you mentioned, and added the pairing timeout that was missing!

**Comment by anastasiyaig** [comment]
@richard-ramos thank you for the fixes! Some more notes from me:

1. when i try to sync a device using expired code -> it results in dead end on the second instance

Screenshot 2023-03-01 at 12 49 06

2. when i try to sync devices with valid non expired code -> then error shown on device 1, device 2 has this white screen with no possibility to proceed

Screenshot 2023-03-01 at 12 50 42

3. when i restart the instance 2 and try to login (after point 2) -> DB migration error is shown
Screenshot 2023-03-01 at 12 51 23

can u please have a look?

**Comment by richard-ramos** [comment]
Ah ... looks like https://github.com/status-im/status-go/pull/3202 introduced a new breaking change. I see a `device type is empty` error in the logs. 

@qfrank, @Samyoul,  what should we send in `DeviceType string 'json:"deviceType"'`?

**Comment by qfrank** [comment]
> Ah ... looks like [status-im/status-go#3202](https://github.com/status-im/status-go/pull/3202) introduced a new breaking change. I see a `device type is empty` error in the logs.
> 
> @qfrank, @Samyoul, what should we send in `DeviceType string 'json:"deviceType"'`?

Hi @richard-ramos , i guess it should be "desktop", 
image

**Comment by igor-sirotin** [comment]
@qfrank @richard-ramos 
As I figured out, `DeviceType` is actually `hostOS`. I noted all possible options here:
https://github.com/status-im/status-desktop/blob/19a9bbcd008e429e51c3a58c8d6548e4f6e9d04c/src/app_service/service/devices/dto/installation.nim#L7-L12

**Comment by Samyoul** [comment]
@igor-sirotin @richard-ramos do you use `DeviceType` as the `hostOS` in desktop? What @qfrank describes looks correct.

Metadata about devices we've paired with use the following struct:
```go
type InstallationMetadata struct {
	// The name of the device
	Name string `json:"name"`
	// The type of device
	DeviceType string `json:"deviceType"`
	// The FCMToken for mobile devices
	FCMToken string `json:"fcmToken"`
}
```

### `Name`

In `status-go` `Name` is `os.Hostname`, unless it has previously been set (this is hard coded in `status-go`):
```go
func (m *Messenger) setInstallationHostname() error {
	imd, err := m.getOurInstallationMetadata()
	if err != nil {
		return err
	}

	// If the name is already set, don't do anything
	if len(imd.Name) != 0 {
		return nil
	}

	hn, err := server.GetDeviceName()
	if err != nil {
		return err
	}
	imd.Name = fmt.Sprintf("%s %s", hn, imd.Name)
	return m.setInstallationMetadata(m.installationID, imd)
}
```

```go
func GetDeviceName() (string, error) {
	name, err := os.Hostname()
	if err != nil {
		return "", err
	}
	return parseHostname(name), nil
}
```

### `DeviceType`

`DeviceType` is `"ios", "android", "desktop"`. Whatever is used is fine, just I think that the UI relies on this field to choose an icon representing the device type, so that we can have consistency across the devices type representations.

**Comment by richard-ramos** [comment]
@Samyoul according to a comment in status-desktop code:
```
# NOTE: DeviceType equals to:
#       - on Desktop (`hostOS` from system.nim):
#           "windows", "macosx", "linux", "netbsd", "freebsd",
#           "openbsd", "solaris", "aix", "haiku", "standalone".
#       - on Mobile (from platform.cljs):
#           "android", "ios"
```

**Comment by anastasiyaig** [comment]
thanks for fixes @richard-ramos !

there is still a crash when i login on second instance when doing sync:

```
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/nim_status_client.nim(180) nim_status_client
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/nim_status_client.nim(177) mainProc
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimqml/src/nimqml/private/qguiapplication.nim(29) exec
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimqml/src/nimqml/private/qobject.nim(49) qobjectCallback
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimqml/src/nimqml/private/qobject.nim(34) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimbus-build-system/vendor/Nim/lib/core/macros.nim(566) onSlotCalled
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/core/signals/signals_manager.nim(56) receiveSignal
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/core/signals/signals_manager.nim(53) processSignal
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/core/eventemitter.nim(57) emit
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app_service/service/devices/service.nim(73) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/core/eventemitter.nim(57) emit
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/modules/main/profile_section/devices/controller.nim(38) :anonymous
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/modules/main/profile_section/devices/io_interface.nim(19) updateOrAddDevice
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/modules/main/profile_section/devices/module.nim(91) updateOrAddDevice
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/src/app/modules/main/profile_section/devices/model.nim(123) updateItem
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimqml/src/nimqml/private/qabstractitemmodel.nim(237) dataChanged
/Users/jenkins/workspace/s-desktop_branches_macos_PR-9482@2/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
```

but i guess this is what @igor-sirotin was mentioning at the very beginning, after instance restart the app works normally. Let me know if this crash should be logged as a separate ticket!

**Review by anastasiyaig — APPROVED** [review]

**Comment by glitchminer** [comment]
Hi @igor-sirotin , any idea what's going on with the behaviour here? I launched the CI build on mac and after entering the sync code am getting stuck on the screen below. It appears there is some element present based on the mouse pointer but it is hidden. 

image

When checking on Windows I got the same thing. This is the console output from the Windows local build (it appears that the Windows CI build is failing).
```
DBG 2023-03-14 01:00:09.414+00:00 primary_action - set state                 topics="app-controller" tid=22352 file=module.nim:186 setCurrFlow=FirstRunOldUserSyncCode setCurrState=SyncDeviceResult
dial tcp 10.5.0.2:57831: connectex: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.
Warning: QMetaObject::invokeMethod: No such method Service::asyncActivityNotificationLoaded(QString) (:0, )
Warning: qrc:/imports/shared/views/DeviceSyncingView.qml:38:5: QML StatusScrollView: Binding loop detected for property "contentHeight" (qrc:/imports/shared/views/DeviceSyncingView.qml:38, )

```

**Comment by igor-sirotin** [comment]
@glitchminer can confirm this, thanks for your catch!
Will fix this, as well as CI build, no idea why it's not working 🙁

**Comment by igor-sirotin** [comment]
@glitchminer that bug appeared after rebasing. 
For now pushed a fix.
And opened a discussion for fair solution here: https://github.com/status-im/status-desktop/issues/9864

**Comment by igor-sirotin** [comment]
Current state:

## ❌ Mobile -> Desktop
1. I get an error. It doesn't work.
```
dial tcp 192.168.1.87:56479: connect: operation timed out
```
image
2. After error, it's not possible to get to previous screen.

## ❌  Desktop -> Mobile

1. Doesn't work for me. Scanning QR code doesn't do anything.
2. On mobile, there's no error report or any UI reaction.
3. Authentication screen 3 words? Are they needed there?

## ✅  Desktop -> Desktop

1. Right after local pairing, the paired device shows 3 devices instead of 2.
It's ok after app restart.
image

# Summary

I think that mobile<->desktop syncing might not work for me because of local wifi.

**Comment by igor-sirotin** [comment]
> @igor-sirotin i was not testing mobile vs desktop integration here. It was not stated to be working. I assumed u implemented only desktop vs desktop syncing, as it was discussed in my first comments to this PR.

@anastasiyaig no worries at all, you're right!
It's just that some fixes landed to mobile since then, so it's partially working already.

I'd prefer to fix it in further PRs, not to block this big one, if it's ok with you 🙂

**Comment by glitchminer** [comment]
Hey @igor-sirotin , I got your contact request message in activity center but I cannot accept the request to reply in app for some reason. 

Just to confirm, if we merge as it is now, then we're saying the backend code will be put in but will be unusable given the current issues. I'm OK with that, but after which PR should we take another look at this?

**Comment by igor-sirotin** [comment]
@glitchminer I will open a series of issues for each problem.
So that we could track that separately and merge PRs faster.

**Comment by glitchminer** [comment]
OK, thanks. Let's do that. We can just keep track on the linked issue.

**Comment by igor-sirotin** [comment]
@glitchminer opened issues for each problem, all tracked here: 
https://github.com/status-im/status-desktop/issues/9895

I've removed some mobile issues from the list, discussed those directly with mobile team.

**Pull request merged**